internal/commands/sbomtest/snykclient/snykclient.go

```diff
@@ -0,0 +1,126 @@
+package snykclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+const sbomTestAPIVersion = "2023-08-31~beta"
+
+type SnykClient struct {
+	http       *http.Client
+	apiBaseURL string
+	orgID      string
+}
+
+func NewSnykClient(httpClient *http.Client, apiBaseURL, orgID string) *SnykClient {
+	return &SnykClient{
+		http:       httpClient,
+		apiBaseURL: apiBaseURL,
+		orgID:      orgID,
+	}
+}
+
+func (s *SnykClient) CreateSBOMTest(ctx context.Context, sbomReader io.Reader) (*SBOMTest, error) {
+	url := fmt.Sprintf("%s/rest/orgs/%s/sbom_tests?version=%s", s.apiBaseURL, s.orgID, sbomTestAPIVersion)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, sbomReader)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/vnd.api+json")
+
+	rsp, err := s.http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	var response CreateSBOMTestRunResponseBody
+	if err := parseResponse(rsp, http.StatusCreated, &response); err != nil {
+		return nil, err
+	}
+
+	return &SBOMTest{
+		Client: s,
+		ID:     response.Data.ID,
+	}, nil
+}
+
+// SBOM Test Client
+
+type SBOMTest struct {
+	Client *SnykClient
+	ID     string
+}
+
+func (s *SBOMTest) GetResult(ctx context.Context) (*GetSBOMTestResultResponseBody, error) {
+	url := s.resultsURL(s.ID)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := s.Client.http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	var response GetSBOMTestResultResponseBody
+	if err := parseResponse(resp, http.StatusOK, &response); err != nil {
+		return nil, err
+	}
+
+	return &response, nil
+}
+
+func (s *SBOMTest) GetStatus(ctx context.Context) (bool, error) {
```

Comment:
i wonder if we should expose this on the test's interface. i think it would be awesome if the consumer would only need to worry about test creation, and didn't have to implement the polling logic.

```go
// where ever we consume this package
c := NewSnykClient()
test, _ := c.CreateSBOMTest()
// the next line would block until the test is complete
// no need for a `for { ... }` loop or anything similar
res, _ := c.GetResult()
``` 

what do you think?

Reply:
I was in two minds about this. On one side it moves it away from being a pure client on the other side it makes the usage of the client much easier.

If we bake the polling logic into the client I think it should take polling configuration into account, such as
- max attempts
- poll frequency / backoff strategy

Alternatively there could be a helper function to take care of this which can then be reused by other aysnc endpoints.

What do you think?

Reply:
I'm personally in favour of taking the burden off the consumer and design it for convenience of use. Exposing the option to set polling parameters does make a lot of sense though. Maybe we can implement the polling in the client for now, and add an interface for the parameters later? At this point I guess we won't have a need to make this available like you would with a shared lib. I'd just avoid that we have to implement the entire polling logic in the callsite of `CreateSBOMTest`, `GetResult`.